### PR TITLE
Fix S1854 FP: Throw should visit finally

### LINKED ISSUE #9466: Fix S1854 FP: Throw should visit finally
We need to add a branch for Throw semantics to connect the blocks
```csharp
    public int ReadInFinallyAfterCatch()
    {
        var value = 0;
        try
        {
            CanThrow();
            value = 42;
        }
        catch
        {
            value = 1;  // Noncompliant FP, used in finally after the throw
            throw;
        }
        finally
        {
            Log(value);
        }
        return value;
    }
```

### PR DESCRIPTION
**Pull request by sebastien-marichal**
Fixes #9466 
Depends on #9528

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.Test/LiveVariableAnalysis/RoslynLiveVariableAnalysisTest.TryCatchFinally.cs [hunk 1]
```diff
@@ -1211,4 +1211,200 @@ public void TryCatchFinally_LiveIn()
         context.Validate("Method(usedAfter);", LiveIn("usedAfter"));
         context.ValidateExit();
     }
+
+    [TestMethod]
+    public void TryCatchFinally_RethrowWithFinally()
```
Comment (mary-georgiou-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/LiveVariableAnalysis/RoslynLiveVariableAnalysisTest.TryCatchFinally.cs:1216]:
```suggestion
    public void TryCatchFinally_Rethrow_ValueLivesOut()
```

File: analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs [hunk 2]
```diff
@@ -174,6 +175,20 @@ private void BuildBranchesNestedCatchRethrow(BasicBlock block)
         }
     }
 
+    private void BuildBranchesRethrowFinally(BasicBlock block)
```
Comment (mary-georgiou-sonarsource) [inline analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs:178]: I think we can merge this method with the method I added for the similar case (catch -> outercatch connection if there's a rethrow) like so:

```csharp
     private void BuildBranchesRethrow(BasicBlock block)
    {
        var reachableHandlerRegions = block.EnclosingRegion(ControlFlowRegionKind.TryAndCatch).NestedRegion(ControlFlowRegionKind.Try).ReachableHandlers();
        var reachableCatchAndFinallyBlocks = reachableHandlerRegions.Where(x => x.FirstBlockOrdinal > block.Ordinal).SelectMany(x => x.Blocks(Cfg));
        foreach (var catchBlock in reachableCatchAndFinallyBlocks.Where(x => x.EnclosingRegion.Kind is ControlFlowRegionKind.Catch))
        {
            AddBranch(block, catchBlock);
        }
        if (reachableCatchAndFinallyBlocks.FirstOrDefault() is { } branch)
        {
            AddBranch(block, branch);
        }
    }   
```

WDYT?

**Review by mary-georgiou-sonarsource — CHANGES_REQUESTED** [review]
Some refactoring

**Review by mary-georgiou-sonarsource — CHANGES_REQUESTED** [review]
As discussed I'm not sure why you get the first or default catch or throw.
Please take a look.

File: analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs [hunk 3]
```diff
@@ -167,13 +167,18 @@ private void BuildBranchesFinally(BasicBlock source, ControlFlowRegion finallyRe
...
     {
-        var reachableHandlers = block.EnclosingRegion(ControlFlowRegionKind.TryAndCatch).NestedRegion(ControlFlowRegionKind.Try).ReachableHandlers();
-        foreach (var catchBlock in reachableHandlers.Where(x => x.Kind == ControlFlowRegionKind.Catch && x.FirstBlockOrdinal > block.Ordinal).SelectMany(x => x.Blocks(Cfg)))
+        var reachableHandlerRegions = block.EnclosingRegion(ControlFlowRegionKind.TryAndCatch).NestedRegion(ControlFlowRegionKind.Try).ReachableHandlers();
+        var reachableCatchAndFinallyBlocks = reachableHandlerRegions.Where(x => x.FirstBlockOrdinal > block.Ordinal).SelectMany(x => x.Blocks(Cfg));
+        foreach (var catchBlock in reachableCatchAndFinallyBlocks.Where(x => x.EnclosingRegion.Kind is ControlFlowRegionKind.Catch))
         {
             AddBranch(block, catchBlock);
         }
+        if (reachableCatchAndFinallyBlocks.Any() && reachableCatchAndFinallyBlocks.First() is { EnclosingRegion.Kind: ControlFlowRegionKind.Finally } finallyBlock)
```
Comment (mary-georgiou-sonarsource) [inline analyzers/src/SonarAnalyzer.CFG/LiveVariableAnalysis/RoslynLiveVariableAnalysis.cs:178]:
```suggestion
        if (reachableCatchAndFinallyBlocks.FirstOrDefault() is { EnclosingRegion.Kind: ControlFlowRegionKind.Finally } finallyBlock)
```

File: analyzers/tests/SonarAnalyzer.Test/LiveVariableAnalysis/RoslynLiveVariableAnalysisTest.TryCatchFinally.cs [hunk 4]
```diff
@@ -65,7 +65,7 @@ public void Using_LiveInUntilTheEnd(string usingStatement, string suffix)
     [TestMethod]
     public void Using_Nested_Block_LiveInUntilTheEnd()
     {
-        var code = """
+        const string code = """
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.Test/LiveVariableAnalysis/RoslynLiveVariableAnalysisTest.TryCatchFinally.cs:68]: `cons string` has no performance benefit over `var`:
https://stackoverflow.com/a/49224799
This creates unnecessary visual noise.

**Review by mary-georgiou-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**